src/toxcore/dht/server/mod.rs

```diff
@@ -246,13 +252,62 @@ impl Server {
     /// Run DHT periodical tasks. Result future will never be completed
     /// successfully.
     pub fn run(self) -> IoFuture<()> {
-        let future = self.clone().run_pings_sending().join3(
+        let future = self.clone().run_pings_sending().join4(
             self.clone().run_onion_key_refresing(),
-            self.run_main_loop()
+            self.clone().run_main_loop(),
+            self.clone().run_recover_network_error()
         ).map(|_| ());
         Box::new(future)
     }
 
+    /// Store bootstap nodes
+    pub fn add_initial_bootstrap(&mut self, pn: PackedNode) {
+        self.initial_bootstrap.push(pn);
+    }
+
+    /// Run initial bootstapping, It continues to sending NodesRequest to bootstrap nodes
+    /// until at least one entry inserted to Kbucket
+    /// and send NodesRequest to nodes in Kbucket if all nodes are discarded.
+    /// All nodes in Kbucket are discarded means that network is not available for a while.
+    fn run_recover_network_error(self) -> IoFuture<()> {
+        let interval = Duration::from_secs(BOOTSTRAP_INTERVAL);
+        let wakeups = Interval::new(Instant::now(), interval);
+        let self_c = self.clone();
+        let future = wakeups
+            .map_err(|e| Error::new(ErrorKind::Other, format!("Bootstrap timer error: {:?}", e)))
+            .take_while(move |_| future::ok(self.is_kbucket_empty() || self.is_all_discarded()))
```

Comment:
> This function, like Iterator::take_while, will take elements from the stream until the predicate resolves to false. Once one element returns false it will always return that the stream is done.

Oops.